Calldata doc (#15533)
ethereum/solidity: docs/types/reference-types.rst

    data location can also be returned from functions, but it is not possible to
    allocate such types.

.. note::
    Arrays and structs with ``calldata`` data location declared in a function body
    or as its return parameters must be assigned before being used or returned.

>     This coding style is a workaround for allowing complex initialization patterns.

Reviewer: This still is not correct.
The compiler enforces a requirement that calldata arrays and structs (declared as function local variables or return parameters) must be initialized before used.
Unfortunately, in certain cases, in which non-trivial control flow is used, the compiler can't detect the initialization.
A common workaround utilized in those cases is to assign the problematic variables to themselves to satisfy the rule enforced by the compiler.

    data location can also be returned from functions, but it is not possible to
    allocate such types.

.. note::

>     Arrays and structs with ``calldata`` data location declared in a function body must be assigned before
>     use. Arrays and structs with ``calldata`` data location as the returned parameters of a function
>     must be assigned before returning. This rule is enforced by the compiler to allow complex initialization patterns.

Reviewer: ```suggestion
    Arrays and structs with ``calldata`` data location declared in a function body
    or as its return parameters must be assigned before being used or returned.
```
Just to avoid repetition.

Reviewer: ```
Arrays and structs declared as `calldata` function local variables or return parameters must be assigned before being used or returned.
```

Maybe it is better phrased.

>     The compiler requires that ``calldata`` arrays and structs (when declared as local function
>     variables or return parameters) are initialized before use. However, in some cases with complex
>     control flow, the compiler may not detect this initialization. A common workaround in such cases
>     is to assign these variables to themselves to fulfill the compiler's requirement.
> 

Reviewer: I think the previous text was better, we just needed to adjust the bit about the workaround.

```suggestion
    Arrays and structs with ``calldata`` data location declared in a function body  
    or as its return parameters must be assigned before being used or returned.
    There are certain cases in which non-trivial control flow is used and the compiler
    can't properly detect the initialization.
    A common workaround in such cases is to assign the affected variable to itself before
    the correct initialization takes place.
```

>     must be assigned before returning. This rule is enforced by the compiler to allow complex initialization patterns.
>     A workaround is assign the variable to itself after declaration.

Reviewer: ```suggestion
    must be assigned before returning. This rule is enforced by the compiler to allow complex initialization patterns.
    A workaround is assign the variable to itself after declaration.
```
It is actually the workaround which allows complex initialization patterns. The rule enforced by the compiler is that you can't use the `calldata` arrays and structs without initialization.

    data location can also be returned from functions, but it is not possible to
    allocate such types.

.. note::

>     Arrays and structs with ``calldata`` data location declared in a function body

Reviewer: ```suggestion
    Arrays and structs with ``calldata`` location declared in a function body
```
